optuna/importance/_fanova/_tree.py

```diff
@@ -299,7 +299,8 @@ def _get_node_children_subspaces(
 
 
 def _get_cardinality(search_spaces: numpy.ndarray) -> float:
-    return numpy.prod(search_spaces[:, 1] - search_spaces[:, 0])
+    c = numpy.prod(search_spaces[:, 1] - search_spaces[:, 0])
+    return c if c > 0 else numpy.prod(search_spaces.shape)
```

Comment:
Could you explain this logic taking the product of the shape?

Reply:
It's a port of original implementation, which handles both categorical features and constants in this code path.
https://github.com/automl/random_forest_run/blob/cc2354ed2622c270d96b88a068ae87990e1da054/include/rfr/util.hpp#L49-L62
I think in case of constants, it's just important that this value is strictly positive (it causes `ZeroDivisionError` down the line). Didn't investigate code paths for categorical features.

Reply:
I see, thanks! So the original implementation includes a code branch to for e.g. categorical paramters return the number of choices as the cardinality. In optuna, we're actually one-hot encoding all categorical parameters (e.g. ["foo", "bar", "baz"] gets converted into three numerical parameters internally with range (0, 1)). I'm not sure but do you think this code has the same effect? (In any case, it might be less relevant if we decide to go with your "second thought" approach)